### PR TITLE
Adds support for primitive alias types and enums

### PR DESCRIPTION
**Pull request by dskatz**
This resolves a bug types that have the following format.

e.g.
```
type ACustomType string

type AStruct struct {
   Member *ACustomType
}
```

This addition to the code generator allows for extending `ACustomType` and provides a builder that accepts the user type and casts to the underlying type.

This PR also adds support for enums.  In most cases were a type is an alias of a primitive (e.g string) it is used to create `const` values within a package, aka enums.

e.g.
```
// +kanopy:builder=true,enum=custom;val
type Member thirdparty.ACustomType

const MemberCustom Member = "custom"
const MemberVal Member = "val"

// WithMember is an autogenerated function
func (o *AStruct) WithMember(in Member) *SomeStruct {
		p := thirdparty.Member(in)
		o.Member = &p
	return o
}
```

### CODE REVIEW

File: pkg/generators/builder/testdata/d/d.go [hunk 1]
```diff
@@ -8,3 +8,6 @@ import (
 type DPolicyRule struct {
 	e.MockPolicyRule
 }
+
+// +kanopy:builder=true
+type AliasType e.AliasToString
```
Comment (yuzhouliu9) [inline pkg/generators/builder/testdata/d/d.go:13]: Should a test be added for the case where an aliased primitive type is not enabled, to check that it is not generated?

File: pkg/generators/snippets/setters_test.go [hunk 2]
```diff
@@ -441,6 +441,31 @@ func TestFuncName(t *testing.T) {
...
+	ctx, err := newTestGeneratorContext()
+	require.NoError(t, err)
+
+	someStruct := newTestType(t, "SomeStruct")
+	root := someStruct
+	member := getMemberFromType(t, someStruct, "SomeStruct", "Alias")
+	want := `// WithAlias is an autogenerated function
+func (o *SomeStruct) WithAlias(in Alias) *SomeStruct {
+		p := b.AliasOfString(in)
+		o.Alias = &p
```
Comment (yuzhouliu9) [inline pkg/generators/snippets/setters_test.go:456]: need to indent 1 tab less

**Comment by whwalter** [comment]
I am not sure what I am doing wrong but when I build this branch and run the generator I just get messages about "skipping primatives" and nothing generated.

Referencing the [jaeger draft](https://github.com/10gen/kanopy/pull/316/files), I don't believe the validating webhook WithFailurePolicy builder is used anywhere and the type seems to duplicate the effort of [FailurePolicyType](https://github.com/10gen/kanopy/blob/master/pkg/builder/k8s/mutating_webhook_configuration.go#L10) in /mutating_webhook_configuration.go. 

I am not sure that you could created the wrapped type in a module without importing the admissionregistration type to do the type assertions for the string.

**Comment by dskatz** [comment]
> I am not sure what I am doing wrong but when I build this branch and run the generator I just get messages about "skipping primatives" and nothing generated.
> 
> Referencing the [jaeger draft](https://github.com/10gen/kanopy/pull/316/files), I don't believe the validating webhook WithFailurePolicy builder is used anywhere and the type seems to duplicate come of the effort of [FailurePolicyType](https://github.com/10gen/kanopy/blob/master/pkg/builder/k8s/mutating_webhook_configuration.go#L10) in /mutating_webhook_configuration.go.
> 
> I am not sure that you could created the wrapped type in a module without importing the admissionregistration type to do the type assertions for the string.

Yes, it will duplicate some things until those are refactored. But lets move that conversation to the other PR.

File: pkg/generators/builder/testdata/d/d.go (hunk 1 above)
Comment (dskatz) [reply pkg/generators/builder/testdata/d/d.go:13]: This is added.

**Review by whwalter — APPROVED** [review]
LGTM. 

code-generator never walks the ast of the imported packages does it? It just gets the name of the embeds and uses that in the template patters. 

Mainly wondering if we already walk the package ast for the k8s stuff we could scan the Package.Constants map for enums  vs having to add them to the tag.

**Comment by dskatz** [comment]
>code-generator never walks the ast of the imported packages does it? It just gets the name of the embeds and uses that in the template patters.
> Mainly wondering if we already walk the package ast for the k8s stuff we could scan the Package.Constants map for enums vs having to add them to the tag.

@whwalter  your assumption is correct.  The code-generator does not walk anything in a project's `go.mod`.  Therefore those constants are unavailable in the structs.

**Pull request merged**